### DISCUSSION TITLE
Schema for names and sources in Parquet

### DISCUSSION THREAD

**Discussion by tschaub** [OvertureMaps/data#327] [Discussion]
I'm curious if there has been discussion about a schema for the Parquet output that more closely matches the schema represented by the yaml in this repo.

For example, in [JSON Schema terms](https://github.com/OvertureMaps/schema/blob/3a5e5b62e43a19ae1d24b9cddb3af149003295f7/schema/defs.yaml#L173-L213), the `names` property is an object with a required `common` property and optional `official`, `alternate`, and `short` properties.

In a Parquet schema, this could be represented by a group `names` node with a required `common` child node and optional `official`, `alternate`, and `short` nodes.  In Arrow, this would would be represented by a struct field.

Currently, it looks like the `names` node in Parquet is a `MAP` annotated group - and that it can contain any keys.

```
  optional group names (MAP) {
    repeated group key_value (MAP) {
      required binary key (STRING);
      optional group value (LIST) {
        repeated group bag {
          optional group array_element (MAP) {
            repeated group key_value (MAP) {
              required binary key (STRING);
              optional binary value (STRING);
            }
          }
        }
      }
    }
  }
```

Instead, a Parquet schema that more closely matched the JSON Schema could look like this:
```
  optional group names {
    required group common (LIST) {
      repeated group list {
        optional group element {
          required binary language (STRING);
          required binary value (STRING);
        }
      }
    }
    optional group official (LIST) {
      repeated group list {
        optional group element {
          required binary language (STRING);
          required binary value (STRING);
        }
      }
    }
    optional group alternate (LIST) {
      repeated group list {
        optional group element {
          required binary language (STRING);
          required binary value (STRING);
        }
      }
    }
    optional group short (LIST) {
      repeated group list {
        optional group element {
          required binary language (STRING);
          required binary value (STRING);
        }
      }
    }
  }
```

(Instead of an arbitrary map with any keys, a struct with required `common` and optional `official`, `alternate`, and `short` nodes.)

The `sources` items could also be represented as structs (with with required `property` & `dataset` and optional `recordId`) instead of maps (with any keys).

**Comment by ibnt1** [comment]
Thanks a lot for the feedback! 

Agree and yes, we are going to remove the use of `MAP` in our parquet schema from all properties, for consistency with the json schema and the other properties, in one of our future releases, as soon as we can.
